### PR TITLE
Add Numpy import guard - fix windows smoke tests

### PR DESCRIPTION
**Pull request by atalman**
Follow up after: https://github.com/pytorch/pytorch/pull/106211

Fixes nightly failure build failure:
https://github.com/pytorch/pytorch/actions/runs/5853088203/job/15871284383

```
(testenv) C:\actions-runner\_work\pytorch\pytorch\builder\windows>python C:\actions-runner\_work\pytorch\pytorch/builder\test_example_code\cnn_smoke.py 
C:\actions-runner\_work\pytorch\pytorch\builder\windows\conda\envs\testenv\lib\site-packages\torch\nn\modules\transformer.py:20: UserWarning: Failed to initialize NumPy: No module named 'numpy' (Triggered internally at C:\cb\pytorch_1000000000000\work\torch\csrc\utils\tensor_numpy.cpp:84.)
  device: torch.device = torch.device(torch._C._get_default_device()),  # torch.device('cpu'),
Traceback (most recent call last):
tensor([0.], device='cuda:0', grad_fn=<ViewBackward0>)
  File "C:\actions-runner\_work\pytorch\pytorch\builder\test_example_code\cnn_smoke.py", line 30, in <module>
    optimizer = optim.SGD(net.parameters(), lr=0.001, momentum=0.1)
  File "C:\actions-runner\_work\pytorch\pytorch\builder\windows\conda\envs\testenv\lib\site-packages\torch\optim\sgd.py", line 26, in __init__
    super().__init__(params, defaults)
  File "C:\actions-runner\_work\pytorch\pytorch\builder\windows\conda\envs\testenv\lib\site-packages\torch\optim\optimizer.py", line 266, in __init__
    self.add_param_group(cast(dict, param_group))
  File "C:\actions-runner\_work\pytorch\pytorch\builder\windows\conda\envs\testenv\lib\site-packages\torch\_compile.py", line 22, in inner
    import torch._dynamo
  File "C:\actions-runner\_work\pytorch\pytorch\builder\windows\conda\envs\testenv\lib\site-packages\torch\_dynamo\__init__.py", line 2, in <module>
    from . import allowed_functions, convert_frame, eval_frame, resume_execution
  File "C:\actions-runner\_work\pytorch\pytorch\builder\windows\conda\envs\testenv\lib\site-packages\torch\_dynamo\allowed_functions.py", line 14, in <module>
    import numpy as np
ModuleNotFoundError: No module named 'numpy'
```

Similar to:
https://github.com/pytorch/pytorch/blob/461c703ee6b82efe9a04d09e57856fee58da375e/torch/utils/data/_utils/__init__.py#L38-L42

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
You need to update the callside that uses `np`.

**Review by albanD — APPROVED** [review]
SGTM if it passes CI

**Review by huydhn — APPROVED** [review]

File: torch/_dynamo/allowed_functions.py [hunk 1]
```diff
@@ -315,4 +321,4 @@ def is_builtin_constant(obj):
 
 
 def is_numpy(obj):
-    return isinstance(obj, np.ndarray) or id(obj) in _numpy_function_ids
+    return HAS_NUMPY and (isinstance(obj, np.ndarray) or id(obj) in _numpy_function_ids)
```
Comment (malfet) [inline torch/_dynamo/allowed_functions.py:324]: Nit: this way you don't need to introduce HAS_NUMPY (it feels a bit odd to define both to be frank)
```suggestion
    return (isinstance(obj, np.ndarray)  if np is not None else False) or id(obj) in _numpy_function_ids
```

**Review by malfet — APPROVED** [review]

**Comment by atalman** [comment]
Closing this - the PR will need to be reverted

**Comment by atalman** [comment]
Converting to draft for now since more changes are required

**Comment by lezcano** [comment]
Oops, sorry for this. What is the current policy for NumPy? I see many testing files that have naked NumPy imports, so is the rule "it's alright to import NumPy on tests, but not on importable files that may be used during a non-test run"?

**Comment by atalman** [comment]
> Oops, sorry for this. What is the current policy for NumPy? I see many testing files that have naked NumPy imports, so is the rule "it's alright to import NumPy on tests, but not on importable files that may be used during a non-test run"?

hi @lezcano We want to keep conditional dependency - same as before the https://github.com/pytorch/pytorch/pull/106211 with try/catch guards and HAS_NUMPY flag where required

**Pull request closed without merging**